File: docs/research/internal-consumer/pre-beta-review-2026-09-17/join_collision.mjs
// Run after building both local TypeScript packages. No workflow invocation.
import { Annotation, StateGraph, START, END } from "../../../../packages/typescript/langgraph/node_modules/@langchain/langgraph/dist/index.js";
import { describe } from "../../../../packages/typescript/langgraph/dist/index.js";
import { validateDocument } from "../../../../packages/typescript/spec/dist/index.js";

const State = Annotation.Root({ values: Annotation({ reducer: (a, b) => a.concat(b), default: () => [] }) });
for (const [label, groups] of [
  ["reversed-sources", [["a", "b"], ["b", "a"]]],
  ["delimiter", [["a+b", "c"], ["a", "b+c"]]],
]) {
  const builder = new StateGraph(State);
  for (const name of [...new Set([...groups.flat(), "sink"])].sort()) {
    builder.addNode(name, () => ({ values: [] }));
    if (name !== "sink") builder.addEdge(START, name);
  }
  for (const sources of groups) builder.addEdge(sources, "sink");
  builder.addEdge("sink", END);
  try {
    const document = await describe(builder.compile(), { strict: true });
    console.log(JSON.stringify({ case: label, joins: document.graphs[0].structure.joins, errors: validateDocument(document) }));
  } catch (error) {
    console.log(JSON.stringify({ case: label, exception: error.name, message: error.message }));
  }
}

File: docs/research/internal-consumer/pre-beta-review-2026-09-17/join_collision.py
"""Minimum legal-node-name join collision; no workflow invocation."""

import json
import operator
from typing import Annotated, TypedDict

from agent_topology.langgraph import describe
from agent_topology.spec import validate_document
from langgraph.graph import END, START, StateGraph


class State(TypedDict):
    values: Annotated[list[str], operator.add]


for label, sources in [
    ("reversed-sources", [["a", "b"], ["b", "a"]]),
    ("delimiter", [["a+b", "c"], ["a", "b+c"]]),
]:
    builder = StateGraph(State)
    for name in sorted({name for group in sources for name in group} | {"sink"}):
        builder.add_node(name, lambda state: {"values": []})
        if name != "sink":
            builder.add_edge(START, name)
    for group in sources:
        builder.add_edge(group, "sink")
    builder.add_edge("sink", END)
    compiled = builder.compile()
    try:
        document = describe(compiled, strict=True)
        print(
            json.dumps(
                {
                    "case": label,
                    "joins": document["graphs"][0]["structure"]["joins"],
                    "errors": validate_document(document),
                }
            )
        )
    except Exception as error:
        print(
            json.dumps(
                {
                    "case": label,
                    "exception": type(error).__name__,
                    "message": str(error),
                }
            )
        )
